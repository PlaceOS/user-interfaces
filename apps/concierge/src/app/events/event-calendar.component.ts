import { Component, Input } from '@angular/core';
import { SettingsService } from '@placeos/common';
import { EventStateService } from './event-state.service';
import { map, shareReplay, take } from 'rxjs/operators';
import { addMonths, addWeeks, format } from 'date-fns';
import { ActivatedRoute, Router } from '@angular/router';

@Component({
    selector: 'event-calendar',
    template: `
        <div class="flex items-center border-y border-base-200 w-full">
            <div class="flex-1 px-2 py-4">
                <span *ngIf="is_today | async" class="text-info text-xs">{{
                    (period | async) === 'week' ? 'This Week' : 'This Month'
                }}</span>
            </div>
            <div class="flex-2 flex items-center justify-center space-x-2">
                <div class="font-medium pl-4">
                    {{ (options | async)?.date | date: 'MMM yyyy' }}
                </div>
                <button icon matRipple (click)="previousPeriod()">
                    <app-icon>chevron_left</app-icon>
                </button>
                <button icon matRipple (click)="nextPeriod()">
                    <app-icon>chevron_right</app-icon>
                </button>
            </div>
            <div class="flex-1 flex items-center justify-end space-x-2 py-2">
                <button
                    btn
                    matRipple
                    class="rounded-3xl"
                    [class.inverse]="(period | async) !== 'week'"
                    (click)="setPeriod('week')"
                >
                    Week
                </button>
                <button
                    btn
                    matRipple
                    class="rounded-3xl"
                    [class.inverse]="(period | async) !== 'month'"
                    (click)="setPeriod('month')"
                >
                    Month
                </button>
            </div>
        </div>
        <div class="w-full h-4"></div>
        <div class="relative flex-1 h-1/2 w-full overflow-auto">
            <event-week-view
                *ngIf="(period | async) !== 'month'; else month_calendar"
            ></event-week-view>
            <ng-template #month_calendar>
                <event-month-view></event-month-view>
            </ng-template>
        </div>
    `,
    styles: [
        `
            :host {
                display: flex;
                flex-direction: column;
                height: 100%;
                width: 100%;
            }
        `,
    ],
})
export class EventCalendarComponent {
    public readonly period = this._state.options.pipe(map((_) => _.period));

    public readonly options = this._state.options;
    public readonly is_today = this.options.pipe(
        map((_) => _.date <= Date.now() && _.end > Date.now())
    );
    public readonly event_list = this._state.event_list;
    public readonly event_day_map = this.event_list.pipe(
        map((list) => {
            const map = {};
            for (const event of list) {
                const date = format(event.date, 'yyyy-MM-dd');
                if (!map[date]) map[date] = [];
                map[date].push(event);
            }
            return map;
        }),
        shareReplay(1)
    );

    public get time_format() {
        return this._settings.time_format;
    }

    constructor(
        private _settings: SettingsService,
        private _state: EventStateService,
        private _router: Router,
        private _route: ActivatedRoute
    ) {}

    public setPeriod(period: 'week' | 'month') {
        this._state.setOptions({ period });
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: { period },
            queryParamsHandling: 'merge',
        });
    }

    public async nextPeriod() {
        const { date, period } = await this.options.pipe(take(1)).toPromise();
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: {
                range:
                    period === 'week'
                        ? addWeeks(date, 1).valueOf()
                        : addMonths(date, 1).valueOf(),
            },
            queryParamsHandling: 'merge',
        });
    }

    public async previousPeriod() {
        const { date, period } = await this.options.pipe(take(1)).toPromise();
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: {
                range:
                    period === 'week'
                        ? addWeeks(date, -1).valueOf()
                        : addMonths(date, -1).valueOf(),
            },
            queryParamsHandling: 'merge',
        });
    }
}
