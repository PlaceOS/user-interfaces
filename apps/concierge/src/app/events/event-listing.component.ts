import { Component } from '@angular/core';
import { SettingsService } from '@placeos/common';
import { EventStateService } from './event-state.service';
import { User } from '@placeos/users';
import { tap } from 'rxjs/operators';

@Component({
    selector: 'event-listing',
    template: `
        <simple-table
            class="min-w-[72rem] w-full block text-sm"
            [data]="event_list"
            empty_message="No Regions"
            [columns]="[
                { key: 'date', name: 'Event', content: event_template },
                {
                    key: 'level',
                    name: 'Level',
                    content: level_template,
                    size: '8rem',
                    sortable: false
                },
                {
                    key: 'room',
                    name: 'Room',
                    content: room_template,
                    size: '12rem',
                    sortable: false
                },
                {
                    key: 'interested',
                    name: 'Interested',
                    content: interested_template,
                    size: '6rem',
                    sortable: false
                },
                {
                    key: 'attending',
                    name: 'Attending',
                    content: attending_template,
                    size: '6rem',
                    sortable: false
                },
                {
                    key: 'state',
                    name: 'Status',
                    content: status_template,
                    size: '8.5rem'
                },
                {
                    key: 'permissions',
                    name: 'Published',
                    content: published_template,
                    size: '6rem',
                    sortable: false
                },
                {
                    key: 'actions',
                    name: ' ',
                    content: actions_template,
                    size: '4rem',
                    sortable: false
                }
            ]"
            [sortable]="true"
        ></simple-table>
        <div class="w-full h-20"></div>
        <ng-template #event_template let-item="row">
            <div class="flex items-center space-x-2 px-3 py-2">
                <div date class="flex flex-col items-center leading-tight w-8">
                    <div
                        month
                        class="text-sm font-medium relative top-0.5 opacity-60"
                    >
                        {{ item.date | date: 'MMM' }}
                    </div>
                    <div day class="text-2xl font-light relative -top-0.5">
                        {{ item.date | date: 'd' }}
                    </div>
                </div>
                <div
                    class="flex items-center justify-center h-12 w-12 rounded overflow-hidden bg-base-200 border border-base-200"
                >
                    <img
                        *ngIf="item.images?.length"
                        auth
                        [source]="item.images[0]"
                        class="min-h-full min-w-full object-cover"
                    />
                </div>
                <div details class="flex flex-col">
                    <div class="text-sm">{{ item.title }}</div>
                    <div class="text-xs opacity-40">
                        {{ item.date | date: 'EEEE, ' + time_format }} &ndash;
                        {{
                            item.date + item.duration * 60 * 1000
                                | date: time_format + ' (z)'
                        }}
                    </div>
                    <div class="text-xs">
                        {{
                            (
                                (item.linked_event?.system_id | space | async)
                                    ?.zones | building
                            )?.address
                        }}
                    </div>
                </div>
            </div>
        </ng-template>
        <ng-template #level_template let-item="row">
            <div class="p-4">
                {{
                    (
                        (item.linked_event?.system_id | space | async)?.zones
                        | level
                    )?.display_name
                }}
                <span *ngIf="!item.linked_event?.system_id" class="opacity-30">
                    No Level
                </span>
            </div>
        </ng-template>
        <ng-template #room_template let-item="row">
            <div class="p-4">
                {{
                    (item.linked_event?.system_id | space | async)?.display_name
                }}
                <span *ngIf="!item.linked_event?.system_id" class="opacity-30">
                    No Room
                </span>
            </div>
        </ng-template>
        <ng-template #interested_template let-item="row">
            <button
                icon
                matRipple
                customTooltip
                [content]="view_attendees"
                class="rounded h-12 w-12 mx-auto"
                [disabled]="!item.attendees?.length"
            >
                {{ item.attendees?.length || 0 }}
            </button>
            <ng-template #view_attendees>
                <div
                    class="relative w-[20rem] h-[28rem] overflow-auto bg-white rounded shadow"
                >
                    <attendee-list
                        [list]="item.attendees"
                        [host]="item.user_email"
                        [hide_close]="true"
                        [custom_title]="'Interested'"
                    ></attendee-list>
                </div>
            </ng-template>
        </ng-template>
        <ng-template #attending_template let-item="row">
            <button
                icon
                matRipple
                customTooltip
                [content]="view_attendees"
                class="rounded h-12 w-12 mx-auto"
                [disabled]="!checkedInCount(item.attendees)"
            >
                {{ checkedInCount(item.attendees) }}
            </button>
            <ng-template #view_attendees>
                <div
                    class="relative w-[20rem] h-[28rem] overflow-auto bg-white rounded shadow"
                >
                    <attendee-list
                        [list]="checkedInList(item.attendees)"
                        [host]="item.user_email"
                        [hide_close]="true"
                    ></attendee-list>
                </div>
            </ng-template>
        </ng-template>
        <ng-template #published_template let-row="row">
            <div
                *ngIf="row.permission === 'OPEN' || row.permission === 'open'"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-success text-success-content mx-auto"
            >
                <app-icon>done</app-icon>
            </div>
        </ng-template>
        <ng-template #status_template let-item="row">
            <div class="p-4">
                <div
                    class="px-4 py-1 rounded-full"
                    [class.bg-success]="
                        item.state !== 'done' &&
                        item.state !== 'in_progress' &&
                        item.state !== 'started'
                    "
                    [class.text-success-content]="
                        item.state !== 'done' &&
                        item.state !== 'in_progress' &&
                        item.state !== 'started'
                    "
                    [class.bg-warning]="
                        item.state === 'in_progress' || item.state === 'started'
                    "
                    [class.text-warning-content]="
                        item.state === 'in_progress' || item.state === 'started'
                    "
                    [class.bg-base-200]="item.state === 'done'"
                    [class.text-base-content]="item.state === 'done'"
                >
                    {{
                        item.state === 'done'
                            ? 'Done'
                            : item.state === 'in_progress' ||
                              item.state === 'started'
                            ? 'In Progress'
                            : 'Active'
                    }}
                </div>
            </div>
        </ng-template>
        <ng-template #actions_template let-row="row">
            <button
                icon
                matRipple
                class="h-12 w-12 rounded mx-2"
                [matMenuTriggerFor]="menu"
                [disabled]="row.state === 'done'"
            >
                <app-icon>more_vert</app-icon>
            </button>
            <mat-menu #menu="matMenu">
                <button mat-menu-item [disabled]="true">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">
                            confirmation_number
                        </app-icon>
                        <div class="mr-2">Promote Event</div>
                    </div>
                </button>
                <button mat-menu-item (click)="viewEvent(row)">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">visibility</app-icon>
                        <div class="mr-2">View Event</div>
                    </div>
                </button>
                <a
                    mat-menu-item
                    [routerLink]="[
                        '/entertainment',
                        'events',
                        'manage',
                        row?.id
                    ]"
                >
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">edit</app-icon>
                        <div class="mr-2">Edit Event</div>
                    </div>
                </a>
                <button mat-menu-item [disabled]="true">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">content_copy</app-icon>
                        <div class="mr-2">Copy URL</div>
                    </div>
                </button>
                <button mat-menu-item (click)="removeEvent(row)">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl text-error">delete</app-icon>
                        <div class="mr-2">Delete Event</div>
                    </div>
                </button>
            </mat-menu>
        </ng-template>
    `,
    styles: [``],
})
export class EventListingComponent {
    public readonly event_list = this._state.event_list.pipe(
        tap((_) => console.log('Event List:', _))
    );

    public readonly viewEvent = (event: any) => this._state.viewEvent(event);
    public readonly removeEvent = (event: any) =>
        this._state.removeEvent(event);

    public get time_format() {
        return this._settings.time_format;
    }

    constructor(
        private _settings: SettingsService,
        private _state: EventStateService
    ) {}

    public checkedInCount(attendees: User[]) {
        if (!attendees?.length) return 0;
        return attendees.filter((user: User) => user.checked_in).length;
    }

    public checkedInList(attendees: User[]) {
        if (!attendees?.length) return [];
        return attendees.filter((user: User) => user.checked_in);
    }
}
