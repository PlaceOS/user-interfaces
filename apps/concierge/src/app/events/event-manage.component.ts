import { COMMA, ENTER, SPACE } from '@angular/cdk/keycodes';
import { Component } from '@angular/core';
import { MatChipInputEvent } from '@angular/material/chips';
import { ActivatedRoute, Router } from '@angular/router';
import {
    BookingFormService,
    removeBooking,
    showBooking,
} from '@placeos/bookings';
import {
    AsyncHandler,
    SettingsService,
    TIMEZONES_IANA,
    currentUser,
    getInvalidFields,
    notifyError,
    randomString,
    unique,
} from '@placeos/common';
import {
    Building,
    BuildingLevel,
    OrganisationService,
} from '@placeos/organisation';
import { first, startWith, take } from 'rxjs/operators';
import { EventStateService } from './event-state.service';
import {
    differenceInMinutes,
    format,
    formatDuration,
    startOfDay,
} from 'date-fns';
import { CalendarEvent, EventFormService, removeEvent } from '@placeos/events';
import { StaffUser, User } from '@placeos/users';

const EMPTY = [];

@Component({
    selector: 'app-event-manage',
    template: `
        <div
            class="absolute inset-0 bg-base-100 overflow-auto"
            *ngIf="!loading; else load_state"
        >
            <a
                icon
                matRipple
                class="sticky top-2 right-2 ml-auto"
                [routerLink]="['/entertainment', 'events']"
            >
                <app-icon>close</app-icon>
            </a>
            <form
                class="flex flex-col w-[40rem] max-w-full mx-auto px-4"
                [formGroup]="form"
            >
                <h3 class="py-4 text-2xl font-medium">
                    {{ form.value.id ? 'Edit' : 'New' }} Group Event
                </h3>
                <section class="flex flex-col space-y-2">
                    <label for="title">Event Title<span>*</span></label>
                    <ng-container>
                        <mat-form-field appearance="outline" class="w-full">
                            <input
                                matInput
                                name="title"
                                formControlName="title"
                                placeholder="e.g. Team Meeting"
                            />
                            <mat-error>
                                {{ 'FORM.TITLE_ERROR' | translate }}
                            </mat-error>
                        </mat-form-field>
                        <label for="host">Organiser<span>*</span></label>
                        <a-user-search-field
                            class="block mb-4"
                            name="host"
                            formControlName="user"
                        ></a-user-search-field>
                        <label for="tags">Tags</label>
                        <mat-form-field
                            appearance="outline"
                            class="no-subscript"
                        >
                            <mat-chip-grid #chipList aria-label="Event Tags">
                                <mat-chip-row
                                    *ngFor="let tag of tag_list"
                                    [selectable]="true"
                                    [removable]="true"
                                    (removed)="removeTag(tag)"
                                >
                                    {{ tag }}
                                    <app-icon matChipRemove>close</app-icon>
                                </mat-chip-row>
                                <input
                                    placeholder="Add new tags relevant to this event..."
                                    [matChipInputFor]="chipList"
                                    [matChipInputSeparatorKeyCodes]="separators"
                                    [matChipInputAddOnBlur]="true"
                                    (matChipInputTokenEnd)="addTag($event)"
                                />
                            </mat-chip-grid>
                        </mat-form-field>
                        <div class="py-4 flex items-center space-x-8">
                            <mat-checkbox formControlName="featured">
                                Featured
                            </mat-checkbox>
                            <mat-checkbox
                                [ngModel]="form.value.permission === 'OPEN'"
                                [ngModelOptions]="{ standalone: true }"
                                (ngModelChange)="
                                    form.patchValue({
                                        permission: $event ? 'OPEN' : 'PRIVATE'
                                    })
                                "
                            >
                                Publish
                            </mat-checkbox>
                        </div>
                    </ng-container>
                    <!-- END BASIC DETAILS -->
                    <div class="w-full h-px bg-base-200"></div>
                    <h3 class="py-4 text-2xl font-medium">Date and Time</h3>
                    <ng-container>
                        <div class="flex items-center flex-wrap sm:space-x-2">
                            <div class="flex-1 min-w-[256px] relative">
                                <label for="date">
                                    {{ 'FORM.DATE' | translate }}<span>*</span>
                                </label>
                                <a-date-field
                                    name="date"
                                    formControlName="date"
                                    [to]="end_date"
                                >
                                    {{ 'FORM.DATE_ERROR' | translate }}
                                </a-date-field>
                            </div>
                            <div class="flex-1 min-w-[256px] relative">
                                <label for="date">
                                    {{ 'FORM.END_DATE' | translate
                                    }}<span>*</span>
                                </label>
                                <a-date-field
                                    name="date"
                                    formControlName="date_end"
                                    [from]="start_date"
                                    [to]="end_date"
                                >
                                    {{ 'FORM.DATE_ERROR' | translate }}
                                </a-date-field>
                            </div>
                        </div>
                        <div
                            class="flex items-center space-x-2"
                            *ngIf="!form.value.all_day"
                        >
                            <div class="flex-1 w-1/3">
                                <label for="start-time">
                                    {{ 'FORM.START_TIME' | translate }}
                                    <span>*</span>
                                </label>
                                <a-time-field
                                    name="start-time"
                                    [ngModel]="form.getRawValue().date"
                                    (ngModelChange)="
                                        form.patchValue({ date: $event })
                                    "
                                    [disabled]="form.controls.date.disabled"
                                    [ngModelOptions]="{ standalone: true }"
                                    [use_24hr]="use_24hr"
                                ></a-time-field>
                            </div>
                            <div class="flex-1 w-1/3">
                                <label for="end-time">
                                    {{ 'FORM.END_TIME' | translate
                                    }}<span>*</span>
                                </label>
                                <a-time-field
                                    name="end-time"
                                    [ngModel]="
                                        form.value.date_end ||
                                        form.value.date +
                                            form.value.duration * 60 * 1000
                                    "
                                    (ngModelChange)="
                                        form.patchValue({ date_end: $event })
                                    "
                                    [ngModelOptions]="{ standalone: true }"
                                    [from]="
                                        form?.getRawValue()?.date +
                                        30 * 60 * 1000
                                    "
                                    [use_24hr]="use_24hr"
                                    [extra_info_fn]="duration_info"
                                ></a-time-field>
                            </div>
                        </div>
                        <div class="flex flex-col">
                            <label for="display-name" i18n="@@displayNameLabel">
                                Timezone:
                            </label>
                            <mat-form-field appearance="outline">
                                <app-icon matPrefix class="text-2xl">
                                    search
                                </app-icon>
                                <input
                                    matInput
                                    formControlName="timezone"
                                    placeholder="Timezone"
                                    [matAutocomplete]="auto"
                                />
                            </mat-form-field>
                            <mat-autocomplete #auto="matAutocomplete">
                                <mat-option
                                    *ngFor="let tz of filtered_timezones"
                                    [value]="tz"
                                >
                                    {{ tz }}
                                </mat-option>
                                <mat-option
                                    *ngIf="!timezones.length"
                                    [disabled]="true"
                                >
                                    No matching timezones
                                </mat-option>
                            </mat-autocomplete>
                        </div>
                    </ng-container>
                    <!-- END DATE TIME -->
                    <div class="w-full h-px bg-base-200"></div>
                    <h3 class="py-4 text-2xl font-medium">Location</h3>
                    <ng-container>
                        <div class="flex items-center space-x-2 pb-2">
                            <button
                                btn
                                matRipple
                                class="flex-1"
                                [class.inverse]="false"
                                [class.inverse]="
                                    form.value.attendance_type !== 'ONSITE'
                                "
                                (click)="
                                    form.patchValue({
                                        attendance_type: 'ONSITE'
                                    })
                                "
                            >
                                <app-icon class="text-2xl">domain</app-icon>
                                <div class="mx-2">On Premise</div>
                            </button>
                            <button
                                btn
                                matRipple
                                class="flex-1"
                                [class.inverse]="
                                    form.value.attendance_type !== 'ONLINE'
                                "
                                (click)="
                                    form.patchValue({
                                        attendance_type: 'ONLINE'
                                    })
                                "
                            >
                                <app-icon class="text-2xl">laptop_mac</app-icon>
                                <div class="mx-2">Online Only</div>
                            </button>
                            <button
                                btn
                                matRipple
                                class="flex-1"
                                [class.inverse]="
                                    form.value.attendance_type !== 'ANY'
                                "
                                (click)="
                                    form.patchValue({ attendance_type: 'ANY' })
                                "
                            >
                                <app-icon class="text-2xl">add</app-icon>
                                <div class="mx-2">Both</div>
                            </button>
                        </div>
                        <ng-container
                            *ngIf="form.value.attendance_type !== 'ONLINE'"
                        >
                            <label for="location">Building Location</label>
                            <mat-form-field appearance="outline">
                                <mat-select
                                    [ngModel]="building_zone"
                                    [ngModelOptions]="{ standalone: true }"
                                    (ngModelChange)="
                                        form.controls.zones.setValue([
                                            $event.parent_id,
                                            $event.id
                                        ]);
                                        setBuilding($event)
                                    "
                                    placeholder="Select Building"
                                >
                                    <mat-option
                                        *ngFor="
                                            let building of building_list
                                                | async
                                        "
                                        [value]="building"
                                    >
                                        {{
                                            building.display_name ||
                                                building.name
                                        }}
                                    </mat-option>
                                </mat-select>
                            </mat-form-field>
                            <div class="flex space-x-2">
                                <div class="flex flex-col flex-[2]">
                                    <label for="level">Level</label>
                                    <mat-form-field appearance="outline">
                                        <mat-select
                                            [ngModel]="level_zone"
                                            [ngModelOptions]="{
                                                standalone: true
                                            }"
                                            (ngModelChange)="setLevel($event)"
                                            placeholder="Select Level"
                                        >
                                            <mat-option
                                                *ngFor="
                                                    let level of active_levels
                                                        | async
                                                "
                                                [value]="level"
                                            >
                                                {{
                                                    level.display_name ||
                                                        level.name
                                                }}
                                            </mat-option>
                                        </mat-select>
                                    </mat-form-field>
                                </div>
                                <div class="flex flex-col flex-[3]">
                                    <label for="level">Room</label>
                                    <mat-form-field appearance="outline">
                                        <mat-select
                                            formControlName="secondary_resource"
                                            [disabled]="
                                                (available_spaces | async)
                                                    ?.length === 0
                                            "
                                            placeholder="Select Room"
                                        >
                                            <mat-option><i>None</i></mat-option>
                                            <mat-option
                                                *ngFor="
                                                    let room of available_spaces
                                                        | async
                                                "
                                                [value]="room.id"
                                            >
                                                {{
                                                    room.display_name ||
                                                        room.name
                                                }}
                                            </mat-option>
                                        </mat-select>
                                    </mat-form-field>
                                </div>
                            </div>
                        </ng-container>
                    </ng-container>
                    <!-- END LOCATION -->
                    <div class="w-full h-px bg-base-200"></div>
                    <h3 class="py-4 text-2xl font-medium">Event Information</h3>
                    <ng-container>
                        <label for="title">Event Description</label>
                        <rich-text-input
                            formControlName="description"
                        ></rich-text-input>
                        <label for="tags">Images</label>
                        <image-list-field
                            formControlName="images"
                        ></image-list-field>
                    </ng-container>
                </section>
                <div
                    class="flex justify-end space-x-2 sticky bottom-0 w-full bg-base-100 py-4"
                >
                    <a
                        btn
                        matRipple
                        class="inverse w-32"
                        [routerLink]="['/entertainment', 'events']"
                    >
                        Cancel
                    </a>
                    <button btn matRipple class="w-32" (click)="save()">
                        Save
                    </button>
                </div>
            </form>
        </div>
        <ng-template #load_state>
            <div
                class="absolute inset-0 bg-base-100 flex flex-col items-center justify-center space-y-4"
            >
                <mat-spinner diameter="48"></mat-spinner>
                <p>Saving event...</p>
            </div>
        </ng-template>
    `,
    styles: [``],
})
export class EventManageComponent extends AsyncHandler {
    public loading = false;
    public timezones: string[] = [];
    public filtered_timezones: string[] = [];

    public readonly form = this._form_state.form;
    public readonly separators: number[] = [ENTER, COMMA, SPACE];
    public readonly building_list = this._org.building_list;
    public readonly active_levels = this._org.active_levels;
    public readonly available_spaces = this._event_form.available_spaces;

    public get tag_list() {
        return this.form.getRawValue().tags || EMPTY;
    }

    public get max_duration() {
        return this._settings.get('app.events.max_duration') || 480;
    }

    public get use_24hr() {
        return this._settings.get('app.use_24_hour_time');
    }

    public get building_zone() {
        const zones = this.form.getRawValue().zones || [];
        const building = this._org.buildings.find((b) => zones.includes(b.id));
        return building || this._org.building;
    }

    public get level_zone() {
        const zones = this.form.getRawValue().zones;
        const level = this._org.levelWithID(zones);
        return level;
    }

    public readonly duration_info = (time: number) => {
        const date = this.form.getRawValue().date;
        if (format(date, 'yyyy-MM-dd') !== format(time, 'yyyy-MM-dd'))
            return '';
        const diff = differenceInMinutes(time, date);
        return ` (${formatDuration({
            hours: Math.floor(diff / 60),
            minutes: diff % 60,
        })})`;
    };

    constructor(
        private _form_state: BookingFormService,
        private _event_form: EventFormService,
        private _state: EventStateService,
        private _route: ActivatedRoute,
        private _router: Router,
        private _org: OrganisationService,
        private _settings: SettingsService
    ) {
        super();
    }

    public async ngOnInit() {
        await this._org.initialised.pipe(first((_) => _)).toPromise();
        this._form_state.setOptions({ type: 'group-event' });
        this.form.patchValue({
            booking_type: 'group-event',
            asset_id: `${
                currentUser()?.name || currentUser()?.email
            } [${randomString(4)}]`,
            permission: 'OPEN',
            zones: unique([
                this._org.organisation.id,
                this._org.building.id,
                this._org.building.parent_id,
            ]),
            location:
                this._org.building.address || this._org.building.display_name,
            user: currentUser(),
            attendance_type: 'ONSITE',
        });
        this.subscription(
            'route.params',
            this._route.paramMap.subscribe(async (params) => {
                if (params.has('id')) {
                    const booking = await showBooking(
                        params.get('id')
                    ).toPromise();
                    if (!booking)
                        return this._router.navigate([
                            '/entertainment',
                            'events',
                        ]);
                    this._form_state.newForm(booking);
                    console.log('Booking:', booking.tags);
                    if (booking.linked_event) {
                        this._event_form.newForm(
                            new CalendarEvent(booking.linked_event)
                        );
                        this.form.patchValue({
                            secondary_resource: booking.linked_event.system_id,
                        });
                    }
                    this.form.patchValue({
                        tags: booking.tags,
                        user: new StaffUser({
                            id: booking.user_id,
                            email: booking.user_email,
                            name: booking.user_name,
                        }),
                    });
                }
            })
        );
        this.subscription(
            'form_change',
            this.form.valueChanges
                .pipe(startWith(this.form.value))
                .subscribe(() => {
                    this._event_form.form.patchValue(this.form.value);
                    this._event_form.setOptions({
                        zone_ids: [
                            this.level_zone?.id || this.building_zone?.id,
                        ],
                    });
                })
        );
        this.subscription(
            'organiser_change',
            this.form.controls.user.valueChanges.subscribe((value) => {
                const name = value?.name || value?.email;
                const value_name =
                    (name ? `${name} [${randomString(4)}]` : '') ||
                    this.form.getRawValue().asset_id;
                this.form.patchValue({ asset_id: value_name });
            })
        );
        this._updateTimezoneList();
        this.subscription(
            'tz-change',
            this.form.valueChanges.subscribe(() => this._updateTimezoneList())
        );
    }

    public setBuilding(bld: Building) {
        this._org.building = bld;
    }

    public setLevel(level: BuildingLevel) {
        const new_zones = unique([
            this._org.organisation.id,
            this._org.building.parent_id,
            this._org.building.id,
            level.id,
        ]);
        this.form.controls.zones.setValue(new_zones);
    }

    /**
     * Add a feature to the list of features for the item
     * @param event Input event
     */
    public addTag(event: MatChipInputEvent): void {
        if (!this.form || !this.form.controls.tags) return;
        const input = event.chipInput.inputElement;
        const value = event.value;
        const feature_list = this.tag_list;
        if ((value || '').trim()) {
            feature_list.push(value);
            this.form.controls.tags.setValue(feature_list);
        }
        if (input) input.value = '';
    }

    /**
     * Remove tag from the list
     * @param existing_tag Feature to remove
     */
    public removeTag(existing_tag: string): void {
        if (!this.form || !this.form.controls.tags) return;
        const tag_list = this.tag_list;
        const index = tag_list.indexOf(existing_tag);

        if (index >= 0) {
            tag_list.splice(index, 1);
            this.form.controls.tags.setValue(tag_list);
        }
    }

    public async save() {
        this.form.markAllAsTouched();
        if (!this.form.valid) {
            return notifyError(
                `Some form fields are invalid. [${getInvalidFields(this.form)}]`
            );
        }
        this.loading = true;
        const booking = this._form_state.booking;
        // Remove Old Event if linked
        if (
            booking?.id &&
            (booking.linked_event || this.form.getRawValue().secondary_resource)
        ) {
            if (booking.linked_event) {
                await removeEvent(booking.linked_event.event_id, {
                    system_id: booking.linked_event.system_id,
                })
                    .toPromise()
                    .catch((e) => {
                        notifyError(e);
                        this.loading = false;
                        throw e;
                    });
            }
            await removeBooking(booking.id)
                .toPromise()
                .catch((e) => {
                    notifyError(e);
                    this.loading = false;
                    throw e;
                });
            this._form_state.form.patchValue({ id: null });
        }
        // Create event if linked to a space
        if (this.form.getRawValue().secondary_resource) {
            const spaces = await this.available_spaces
                .pipe(take(1))
                .toPromise();
            const space = spaces.find(
                (s) => s.id === this.form.getRawValue().secondary_resource
            );
            const value = { ...this.form.getRawValue() };
            delete value.id;
            console.log('Event value:', value);
            this._event_form.form.patchValue({
                ...value,
                id: null,
                resources: [space],
            });
            const event = await this._event_form.postForm().catch((e) => {
                notifyError(e);
                this.loading = false;
                throw e;
            });
            console.log('Event:', event);
            if (!event) return;
            this.form.patchValue({
                ical_uid: event.ical_uid,
                event_id: event.id,
                location: space.display_name || space.name,
            });
        }
        this.form.patchValue({ asset_name: '' });
        if (!this.form.getRawValue().description)
            this.form.patchValue({ description: ' ' });
        const date = this.form.getRawValue().date;
        const res = await this._form_state.postForm().catch(async (e) => {
            notifyError(e);
            if (this.form.getRawValue().event_id) {
                await removeEvent(this.form.getRawValue().event_id).toPromise();
            }
        });
        this._state.changed();
        this.loading = false;
        if (res) {
            this._router.navigate(['/entertainment', 'events'], {
                queryParams: { range: startOfDay(date).valueOf() },
            });
        }
    }

    private _updateTimezoneList() {
        const timezone = this.form?.value?.timezone || '';
        this.timezones = TIMEZONES_IANA;
        this.filtered_timezones = this.timezones.filter((_) =>
            _.toLowerCase().includes(timezone.toLowerCase())
        );
    }
}
