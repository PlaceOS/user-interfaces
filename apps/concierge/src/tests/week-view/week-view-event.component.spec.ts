import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { CalendarEvent } from '@placeos/events';
import { EventsStateService } from '../../app/day-view/events-state.service';
import { WeekViewEventComponent } from '../../app/week-view/week-view-event.component';

describe('WeekViewEventComponent', () => {
    let spectator: Spectator<WeekViewEventComponent>;
    const createComponent = createComponentFactory({
        component: WeekViewEventComponent,
        providers: [
            { provide: EventsStateService, useValue: { setEvent: jest.fn() } },
        ],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
        spectator.setInput({
            event: new CalendarEvent({ title: 'test', date: 2 }),
        });
        spectator.detectChanges();
        expect(spectator.element).toMatchSnapshot();
    });

    it('should allow viewing more event details', () => {
        const service = spectator.inject(EventsStateService);
        expect(service.setEvent).not.toBeCalled();
        spectator.setInput({
            event: new CalendarEvent({ title: 'test', date: 2 }),
        });
        spectator.detectChanges();
        spectator.click('[content]');
        expect(service.setEvent).toBeCalled();
    });
});
