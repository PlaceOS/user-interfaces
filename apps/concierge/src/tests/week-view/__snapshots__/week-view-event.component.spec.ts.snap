// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekViewEventComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="bg-white border-2 font-normal m-2 null p-2 rounded shadow text-black text-left"
    content=""
    matripple=""
  >
    <div
      class="text-base"
    >
      
    </div>
    <div
      class="text-sm"
    >
      
    </div>
    <div
      class="text-sm"
    >
        Attendees 
    </div>
  </div>
</div>
`;

exports[`WeekViewEventComponent should match snapshot 2`] = `
<div
  id="root1"
>
  <div
    class="bg-white border-2 font-normal m-2 p-2 rounded shadow text-black text-left done"
    content=""
    matripple=""
  >
    <div
      class="text-base"
    >
      test
    </div>
    <div
      class="text-sm"
    >
      10:00 AM
    </div>
    <div
      class="text-sm"
    >
       0 Attendees 
    </div>
  </div>
</div>
`;
